features/268-next-cf/api.md

```diff
@@ -0,0 +1,305 @@
+
+src: https://github.com/openfaas/faas-netes/tree/master/chart/openfaas
+
+Some key configuration:
+- external services would NOT be exposed
+  - skygear admin may need to connect to the gateway through port forwarding
+  - thus gateway basic auth may be optional
+- imagePullPolicy => IfNotPresent, to optimise for function reboot
+  - when the cluster grow, we may need to consider how to further optimise for function boot time
+- scaleFromZero => true, faasIdler.dryRun => false, to scale down function to zero when needed
+- faasIdler.inactivityDuration affects when functions would be scale down to zero, however we will have mainly two types of deployed function, short-lived functions and long-running micro-services, scaling strategy would be discussed in later sections
+
+#### Kubernetes CRD
+
+src: https://github.com/openfaas/faas-netes/tree/master/chart/openfaas#openfaas-operator--crd-controller
+
+CRD may be considered in the future, but as suggested the more reliable faas-netes controller would be used.
+
+## CF Controller
+
+This provides APIs for skygear users to manage functions under their own authority.
+
+CF Controller would delegate the operations by accessing OpenFaas Gateway directly.
+
+Here is the list of APIs:
+- List all deployed functions
+- Get configuration, status and metrics of a function
+- Create / deploy new function
+- Update the configration of a function
+  - This will trigger new deployment of the function
+- Remove a function
+- View log of a function
+- Set function version (explain in later section)
+- Add secrets
+
+CF Controller is multi-tenant. Skygear users must login to skycli beforehead.
+
+## CF Gateway
+
+CF Gateway receives request from Gateway, and
+- forward the request to corresponding function in OpenFaas
+- for configurations of a function would be carried out in the CF Gateway, e.g.
+  - access control by keys, roles, etc, the CF Gateway connect to auth gear or the db directly to do authz
+  - secrets injection (or use k8s secrets? can be discussed)
+
+### Routing to OpenFaas
+
+Steps for a request to reach OpenFaas:
+- Gateway receive http request
+  - if that path start with one that belongs to a gear, it forwards the request to the gear
+  - otherwise it forwards the request to CF Gateway
+- CF Gateway match the path against the function routes stored in CF Config Store
+  - if a function is matched, it rewrite the path to that function and forward the request to the function
+  - otherwise it return function not found error
+
+When Gateway receive a function not found error from CF Gateway, it should transform to a generic route not found error. If the Gateway is deployed without CF Gateway configured, it should immediately return route not found error if path does not match any gear.
+
+Since CF Gateway sits behined the Gateway, there would be a naming restriction of CF routes that it cannot collide the with gear path. So currently CF routes cannot have the prefix `/auth/`.
```

Comment:
I think this "naming restriction" should be careful, which it implies:

1. user can't have cloud function path `/auth/` to overwrite auth gear's path (a user doesn't want to use auth gear but want to have `/auth/`), or extend auth gear's capabilities by adding CF.
2. we should reserve gear names (chat, cms, ...?) for future possible usage.

Reply:
Yeah.

I think we may make some rules so that cf path would not conflict with the gear one, e.g. cf path cannot start with `_` and gear path would all start with `_`.